File: apps/mobile-client/android/app/src/main/java/com/anonymous/mobileclient/WabridgeModule.kt
package com.anonymous.mobileclient

import com.facebook.react.bridge.*
import wabridge.Bridge
import wabridge.MessageCallback
import wabridge.Wabridge

/**
 * WabridgeModule — React Native NativeModule that wraps the Go wabridge .aar.
 *
 * The wabridge.aar is compiled via gomobile bind from packages/wabridge/.
 * Gomobile-generated Java API throws exceptions instead of using error output params.
 */
class WabridgeModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = "Wabridge"

    private var bridge: Bridge? = null

    // ── initBridge ──────────────────────────────────────────────────────────

    @ReactMethod
    fun initBridge(dbPath: String, authToken: String, promise: Promise) {
        runOnBackground(promise) {
            bridge = Bridge(dbPath, authToken)
            null // resolve with null (void)
        }
    }

    // ── startPairing ────────────────────────────────────────────────────────

    @ReactMethod
    fun startPairing(phoneNumber: String, promise: Promise) {
        runOnBackground(promise) {
            val code = requireBridge().startPairing(phoneNumber)
            code
        }
    }

    // ── isLinked ─────────────────────────────────────────────────────────────

    @ReactMethod
    fun isLinked(promise: Promise) {
        runOnBackground(promise) {
            requireBridge().isLinked
        }
    }

    // ── syncAndTriage ────────────────────────────────────────────────────────

    @ReactMethod
    fun syncAndTriage(lastSyncTimestamp: Double, promise: Promise) {
        runOnBackground(promise) {
            val b = requireBridge()
            val callback = object : MessageCallback {
                override fun onMessage(jsonPayload: String) {
                    // Emit to JS via event emitter (optional; JS can poll getMatches)
                }
            }
            val result = b.syncAndTriage(lastSyncTimestamp.toLong(), callback)
            val map = Arguments.createMap()
            map.putInt("messagesSynced", result.messagesSynced.toInt())
            map.putString("error", result.error ?: "")
            map
        }
    }

    // ── getFilters ───────────────────────────────────────────────────────────

    @ReactMethod
    fun getFilters(promise: Promise) {
        runOnBackground(promise) {
            requireBridge().getFilters()
        }
    }

    // ── saveFilter ───────────────────────────────────────────────────────────

    @ReactMethod
    fun saveFilter(filterJson: String, promise: Promise) {
        runOnBackground(promise) {
            requireBridge().saveFilter(filterJson)
        }
    }

    // ── deleteFilter ─────────────────────────────────────────────────────────

    @ReactMethod
    fun deleteFilter(id: String, promise: Promise) {
        runOnBackground(promise) {
            requireBridge().deleteFilter(id)
            null
        }
    }

    // ── getMatches ───────────────────────────────────────────────────────────

    @ReactMethod
    fun getMatches(filterId: String, limit: Int, promise: Promise) {
        runOnBackground(promise) {
            requireBridge().getMatches(filterId, limit.toLong())
        }
    }

    // ── unlink ───────────────────────────────────────────────────────────────

    @ReactMethod
    fun unlink(promise: Promise) {
        runOnBackground(promise) {
            requireBridge().unlink()
            bridge = null
            null
        }
    }

    // ── helpers ──────────────────────────────────────────────────────────────

    private fun requireBridge(): Bridge =
        bridge ?: throw RuntimeException(
            "Wabridge not initialized. Call initBridge(dbPath, authToken) first."
        )

    /**
     * Runs [block] on a background thread and resolves/rejects [promise].
     * Gomobile-generated methods throw exceptions on error.
     */
    private fun runOnBackground(promise: Promise, block: () -> Any?) {
        Thread {
            try {
                val result = block()
                when (result) {
                    null -> promise.resolve(null)
                    is WritableMap -> promise.resolve(result)
                    is String -> promise.resolve(result)
                    is Boolean -> promise.resolve(result)
                    is Int -> promise.resolve(result)
                    is Double -> promise.resolve(result)
                    else -> promise.resolve(result.toString())
                }
            } catch (e: Exception) {
                promise.reject("WABRIDGE_ERROR", e.message ?: "Unknown error", e)
            }
        }.start()
    }
}

File: apps/mobile-client/ios/WabridgeModule.swift
import Foundation
import Wabridge

/**
 * WabridgeModule — React Native NativeModule wrapping the Go Wabridge.xcframework.
 *
 * The xcframework is compiled via:
 *   gomobile bind -target=ios -o ios/Wabridge.xcframework .
 * from packages/wabridge/.
 *
 * Gomobile's ObjC API uses NSError** for errors; Swift bridges these as throws.
 */
@objc(Wabridge)
class WabridgeModule: NSObject {

  private var bridge: WabridgeBridge?

  // MARK: - initBridge

  @objc func initBridge(
    _ dbPath: String,
    authToken: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        var error: NSError?
        guard let b = WabridgeNewBridge(dbPath, authToken, &error) else {
          throw error ?? NSError(domain: "Wabridge", code: -1,
            userInfo: [NSLocalizedDescriptionKey: "Failed to init bridge"])
        }
        self.bridge = b
        resolve(nil)
      } catch {
        reject("WABRIDGE_ERROR", error.localizedDescription, error)
      }
    }
  }

  // MARK: - startPairing

  @objc func startPairing(
    _ phoneNumber: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        guard let b = self.bridge else {
          throw Self.notInitializedError()
        }
        let code = try b.startPairing(phoneNumber)
        resolve(code)
      } catch {
        reject("WABRIDGE_ERROR", error.localizedDescription, error)
      }
    }
  }

  // MARK: - isLinked

  @objc func isLinked(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      guard let b = self.bridge else {
        resolve(false)
        return
      }
      resolve(b.isLinked())
    }
  }

  // MARK: - syncAndTriage

  @objc func syncAndTriage(
    _ lastSyncTimestamp: Double,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .background).async {
      do {
        guard let b = self.bridge else { throw Self.notInitializedError() }

        let callback = MessageCallbackImpl { _ in /* JS polls getMatches */ }
        guard let result = try b.syncAndTriage(Int64(lastSyncTimestamp), callback: callback) else {
          resolve(["messagesSynced": 0, "error": ""])
          return
        }
        resolve(["messagesSynced": result.messagesSynced, "error": result.error])
      } catch {
        reject("WABRIDGE_ERROR", error.localizedDescription, error)
      }
    }
  }

  // MARK: - getFilters

  @objc func getFilters(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        guard let b = self.bridge else { throw Self.notInitializedError() }
        let json = try b.getFilters()
        resolve(json)
      } catch {
        reject("WABRIDGE_ERROR", error.localizedDescription, error)
      }
    }
  }

  // MARK: - saveFilter

  @objc func saveFilter(
    _ filterJson: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        guard let b = self.bridge else { throw Self.notInitializedError() }
        let json = try b.saveFilter(filterJson)
        resolve(json)
      } catch {
        reject("WABRIDGE_ERROR", error.localizedDescription, error)
      }
    }
  }

  // MARK: - deleteFilter

  @objc func deleteFilter(
    _ id: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        guard let b = self.bridge else { throw Self.notInitializedError() }
        try b.deleteFilter(id)
        resolve(nil)
      } catch {
        reject("WABRIDGE_ERROR", error.localizedDescription, error)
      }
    }
  }

  // MARK: - getMatches

  @objc func getMatches(
    _ filterId: String,
    limit: Int,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        guard let b = self.bridge else { throw Self.notInitializedError() }
        let json = try b.getMatches(filterId, limit: Int(limit))
        resolve(json)
      } catch {
        reject("WABRIDGE_ERROR", error.localizedDescription, error)
      }
    }
  }

  // MARK: - unlink

  @objc func unlink(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        guard let b = self.bridge else { throw Self.notInitializedError() }
        try b.unlink()
        self.bridge = nil
        resolve(nil)
      } catch {
        reject("WABRIDGE_ERROR", error.localizedDescription, error)
      }
    }
  }

  // MARK: - helpers

  static func notInitializedError() -> NSError {
    NSError(domain: "Wabridge", code: -1, userInfo: [
      NSLocalizedDescriptionKey: "Wabridge not initialized. Call initBridge(dbPath, authToken) first."
    ])
  }
}

// MARK: - MessageCallback

/// Concrete implementation of WabridgeMessageCallback protocol for gomobile.
private class MessageCallbackImpl: NSObject, WabridgeMessageCallbackProtocol {
  private let handler: (String) -> Void
  init(_ handler: @escaping (String) -> Void) { self.handler = handler }
  func onMessage(_ jsonPayload: String?) { handler(jsonPayload ?? "") }
}
